qkit/drivers/ZHInst_Virtual_VNA.py

```diff
@@ -0,0 +1,132 @@
+        self.integration_length = 4096
+
+    def get_freqpoints(self) -> np.ndarray:
+        return self._frequencies
+
+    def get_tracedata(self, RealImag = None) -> tuple((np.ndarray, np.ndarray)):
+        if not RealImag:
+            signal = (self._i_q_data[:, 0] + 1.0j * self._i_q_data[:, 1])
+            return np.abs(signal), np.angle(signal)
+        else:
+            return self._i_q_data[:, 0], self._i_q_data[:, 1]
+
+    def _get_shfsg_frequencies(self) -> tuple((float, float)):
+        """
+        Returns (SHF-Upconversion center frequency, Digital Modulation Frequency)
+        """
+        shfsg_out_freq = self.get_freqpoints()[0] - MINIMUM_UHFQA_FREQUENCY
+        base_mod = min(int(shfsg_out_freq / 0.5e9) * 0.5e9, 8.0e9) # The maximum CF for the SHFSG
+        return base_mod, shfsg_out_freq - base_mod
+
+    def pre_measurement(self):
+        print("Setup")
+        # Configure output
+        base, digital = self._get_shfsg_frequencies()
+        self.shfsg.set_sgchannels_7_sgchannels_output_range(0)
+        self.shfsg.set_synthesizers_3_synthesizers_centerfreq(base)
+        self.shfsg.set_sgchannels_7_sgchannels_awg_modulation_enable(1)
+        self.shfsg.set_sgchannels_7_sgchannels_oscs_0_freq(digital)
+
+        # Configure upper side band modulation
+        self.shfsg.set_sgchannels_7_sgchannels_sine_oscselect(0)
+        self.shfsg.set_sgchannels_7_sgchannels_sine_i_enable(1)
+        self.shfsg.set_sgchannels_7_sgchannels_sine_i_sin_amplitude(0.0)
+        self.shfsg.set_sgchannels_7_sgchannels_sine_i_cos_amplitude(0.9)
+        self.shfsg.set_sgchannels_7_sgchannels_sine_q_enable(1)
+        self.shfsg.set_sgchannels_7_sgchannels_sine_q_sin_amplitude(0.9)
+        self.shfsg.set_sgchannels_7_sgchannels_sine_q_cos_amplitude(0.0)
+
+        self.uhfqa.compile_program(
+            sequence_type=SequenceType.PULSED_SPEC,
+            period=self.period,
+            repetitions=self.repetitions,
+            trigger_mode=TriggerMode.NONE,
+            alignment=Alignment.START_WITH_TRIGGER
+        )
+        self.shfsg.set_sgchannels_7_sgchannels_output_on(1)
+        self.uhfqa.set_sigouts_0_sigouts_on(1) # Enable the outputs
+        self.uhfqa.set_sigouts_1_sigouts_on(1)
+
+        self.uhfqa.set_outputs_1_awg_outputs_amplitude(-1.0)
+
+        self._i_q_data = np.zeros(shape=(len(self.get_freqpoints()), 2), dtype=complex)
+
+    def start_measurement(self):
+        print("Run")
+        self.uhfqa.arm(length=self.repetitions, averages=1)
+
+        base, digital = self._get_shfsg_frequencies()
+        shfsg_frequency = base + digital
+        uhfqa_frequencies = self.get_freqpoints() - shfsg_frequency
+        assert np.all(uhfqa_frequencies < 600e9), "Exceeding maximum UHFQA frequency!"
+        
+        for i, f in enumerate(uhfqa_frequencies):
+            self.uhfqa.set_osc_freq(f)                            # set modulation frequency
+            self.uhfqa.set_readout_frequency0(f)
```

Comment:
This call costs time and might be unnecessary. In spectroscopy mode, the signal is (at least according to the documentation) automatically demodulated with the local oscillator.